File: DMOJ Python/fibonacci/Fibonacci.py
MOD = 1000000007


def fib(n):
    if (n == 0):
        return (0, 1)
    res = fib(n // 2)
    a = res[0] * (res[1] * 2 - res[0] + MOD) % MOD
    b = ((res[0] * res[0]) % MOD + (res[1] * res[1]) % MOD) % MOD
    if (n % 2 == 0):
        return (a, b)
    return (b, (a + b) % MOD)


N = int(input())

print(fib(N)[0])

File: DMOJ C++/fibonacci/Fibonacci.cpp
#include <bits/stdc++.h>

#define MOD 1000000007ULL

using namespace std;

typedef unsigned long long ull;
typedef pair<ull, ull> pll;

pll fib (ull n) {
    if (n == 0) return make_pair(0ULL, 1ULL);
    pll res = fib(n / 2);
    ull a = res.first * (res.second * 2ULL - res.first + MOD) % MOD;
    ull b = ((res.first * res.first) % MOD + (res.second * res.second) % MOD) % MOD;
    if (n % 2 == 0) return make_pair(a, b);
    return make_pair(b, (a + b) % MOD);
}

int main() {
    ull n;
    scanf("%llu", &n);
    printf("%llu\n", fib(n).first);
}
